cvat-data/src/js/cvat-data.js

```diff
@@ -116,56 +119,96 @@ class FrameProvider {
         if (frameNumber in this._frames) {
             return this._frames[frameNumber];
         }
-
         return null;
     }
 
+
     /*
-        Method decodes a block of data
+        Method start asynchronic decode a block of data
 
         @param block - is a data from a server as is (ts file or archive)
         @param start {number} - is the first frame of a block
         @param end {number} - is the last frame of a block + 1
+        @param needed_frame {number} - number of frame that should be returned as soon as
+        @param callback - callback)
+
     */
-    decode(block, start, end) {
-        async function decode(resolve, reject) {
-            // no more requests possible during one is running
+
+    start_decode(block, start, end, needed_frame, callback)
+    {
+         if (this._blockType === BlockType.TSVIDEO){
             if (this._running) {
                 throw new Error('Decoding has already running');
             }
-
-            try {
-                this._running = true;
-                this._blocks.push(`${start}:${end}`);
-
-                this._cleanup();
-
-                if (this._blockType === BlockType.TSVIDEO) {
-                    this._demuxer.write(block);
-                    for (let i = start; i < end; i++) {
-                        const result = this._videoDecoder.decode();
-                        if (!Array.isArray(result)) {
-                            const message = 'Result must be an array.'
-                                + `Got ${result}. Possible reasons: `
-                                + 'bad video file, unpached jsmpeg';
-                            throw Error(message);
-                        }
-                        this._frames[i] = YCbCrToRGBA(...result);
-                    }
-                    resolve();
-                } else if (this._blockType === BlockType.ARCHIVE) {
-                    // to do with any npm dearchiver
-                }
-            } catch (error) {
-                reject(error);
-            } finally {
-                this._running = false;
+            this._demuxer.write(block);
+            this._blocks.push(`${start}:${end}`);
+            this._cleanup();
+            this._currFrame = start;
+            for (var i = start; i < end; i++)
+            {
+                this._frames[i] = "loading";
             }
-        }
+            this._running = true;
+            this._decode_id = setInterval(this.decode.bind(this), this, start, end, needed_frame, callback, 10);
```

Comment:
- camelCase
- Why do you pass an additional "this" argument after you had binded decode to this?

Reply:
About the second, I mean you can do something like this:
setInterval(this.decode.bind(this, start, end, needed_frame, callback), 10);

Reply:
I would prefer setTimeout instead of setInterval because there can be some difficulties if frame decodings takes more than 10 milliseconds.